ngmt/modules/td/_pham.py

```diff
@@ -0,0 +1,372 @@
+            which is a pandas DataFrame in BIDS format with the following information:
+                - onset: Start time of the turn [s].
+                - duration: Duration of the turn [s].
+                - event_type: Type of the event (turn).
+                - direction_of_turn: Direction of turn which is either left or right.
+                - angle_of_turn_deg: Angle of the turn in degrees.
+                - peak_angular_velocity: Peak angular velocity [deg/s].
+                - tracking_systems: Tracking systems used (default is 'imu').
+                - tracked_points: Tracked points on the body (default is 'LowerBack').
+        """
+        # Check if data is a DataFrame
+        if not isinstance(data, pd.DataFrame):
+            raise ValueError("Input data must be a pandas DataFrame")
+
+        # Check if data has the correct shape
+        if data.shape[1] != 6:
+            raise ValueError("Input data must have 6 columns (3 for accelerometer and 3 for gyro)")
+
+        # Check if sampling frequency is positive
+        if sampling_freq_Hz <= 0:
+            raise ValueError("Sampling frequency must be positive")
+
+        # Check if plot_results is a boolean
+        if not isinstance(plot_results, bool):
+            raise ValueError("plot_results must be a boolean value")
+
+        # Select acceleration data and convert it to numpy array format
+        accel = data.iloc[:, 0:3].copy()
+        accel = accel.to_numpy()
+
+        # Select gyro data and convert it to numpy array format
+        gyro = data.iloc[:, 3:6].copy()
+        gyro = gyro.to_numpy()
+
+        # Check unit of acceleration data if it is in g or m/s^2
+        if acceleration_data_unit == "m/s^2":
+            # Convert acceleration data from m/s^2 to g (if not already is in g)
+            accel /= 9.81
+
+        # Check unit of acceleration data if it is in g or m/s^2
+        if gyro_data_unit == "rad/s":
+            # Convert acceleration data from rad/s to deg/s (if not already is in deg/s)
+            gyro = np.rad2deg(gyro)
+
+        # Convert gyro data unit from deg/s to rad/s
+        gyro *= self.gyro_convert_unit
+
+        # Compute the variance of the moving window of gyro signal
+        gyro_vars = []
+
+        for i in range(3):
+            gyro_var = preprocessing.moving_var(data=gyro[:, i], window=sampling_freq_Hz)
+            gyro_vars.append(gyro_var)
+
+        gyro_var_1, gyro_var_2, gyro_var_3 = gyro_vars
+
+        # Threshold value for identifying periods where the variance is low
+        thr = 2 * 10**-4
+        
+        # Identify periods where the variance is below the threshold
+        self.locs = (gyro_var_1 <= thr) & (gyro_var_2 <= thr) & (gyro_var_3 <= thr)
+        
+        # Exclude the last 'sampling_freq_Hz' samples from the identified periods
+        self.locs[-sampling_freq_Hz:] = False
+        
+        # If no bias periods are found, return the original gyro values
```

Comment:
What are bias periods?

Reply:
The bias periods are intervals where the variance of the gyroscope signal is low, suggesting that the IMU sensor is not experiencing significant rotational movement. This bias is subtracted from the gyroscope data to ensure that the turn detection algorithm works with an unbiased signal.

Here's additional comments to clarify these steps:

1. Identify Bias Periods: Calculate the variance of the gyroscope data over a moving window and identify periods where the variance is below a defined threshold.

2. Exclude Edge Samples: Exclude the last sampling_freq_Hz samples from the identified periods to avoid edge effects in the data.

3. Compute and Subtract Bias: Compute the mean bias from the identified low-variance periods and subtract this bias from the original gyroscope signal.